source/tests/system/enumerate_devices.cpp

```diff
@@ -0,0 +1,39 @@
+#include "enumerate_devices.h"
+
+#include <stdexcept>
+
+#include <server/session_utilities_service.h>
+#include "device_server.h"
+
+namespace ni {
+namespace tests {
+namespace system {
+
+const google::protobuf::RepeatedPtrField<nidevice_grpc::DeviceProperties> EnumerateDevices(bool clear_cache)
+{
+  static bool devices_are_cached = false;
+  static google::protobuf::RepeatedPtrField<nidevice_grpc::DeviceProperties> devices_cache;
+
+  if (!devices_are_cached || clear_cache) {
+    devices_are_cached = false;
+
+    ::nidevice_grpc::SessionUtilities::Stub stub(DeviceServerInterface::Singleton()->InProcessChannel());
+
+    nidevice_grpc::EnumerateDevicesRequest request;
+    nidevice_grpc::EnumerateDevicesResponse response;
+    ::grpc::ClientContext context;
+
+    if (::grpc::StatusCode::OK != stub.EnumerateDevices(&context, request, &response).error_code()) {
+      throw std::runtime_error("Failed to enumerate devices");
+    }
+
+    devices_cache = response.devices();

```

Comment:
Are you sure the `RepeatedPtrField` doesn't get cleaned up when the response message gets freed? Or is this making a copy?

Reply:
I'm not sure it made a copy. So, I updated the code to explicitly copy the `RepeatedPtrField`.
```c++
devices_cache.CopyFrom(response.devices());
```